File: quickcode/frontend/js/toon.js
// TOON in the browser: the same encoding the model reads, so the transcript
// shows what was actually sent rather than a JSON re-render of it.
//
// This mirrors `quickcode/context/toon.py`. Two implementations of one format
// is a cost, and it is paid on purpose: the server never ships tool arguments
// as text (they arrive as the model's JSON, which is the wire format), so the
// only place a TOON view of them can be built is here. Keep the two in step --
// `tests/test_toon.py` pins the Python side, and the shapes below are the ones
// it pins.
//
// INTEGRATION
//   import { toon, highlightToon } from "./toon.js";
//   pre.innerHTML = highlightToon(toon(argsObject));

const COMMA = ",";
export const TAB = "\t";

const INDENT = 2;
const MAX_DEPTH = 24;

const LITERALS = new Set(["true", "false", "null"]);
const NUMBERISH = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$/;
// eslint-disable-next-line no-control-regex
const CONTROL = /[\x00-\x1f\x7f]/;
const KEY_BREAKERS = /[:"{}[\],]/;

const ESCAPES = { "\\": "\\\\", '"': '\\"', "\n": "\\n", "\r": "\\r", "\t": "\\t" };

export function toon(value, { delimiter = COMMA, indent = INDENT } = {}) {
  const lines = [];
  write(null, value, lines, 0, delimiter, indent, 0);
  return lines.join("\n");
}

// ---- the four forms ----

function write(key, value, lines, depth, d, indent, guard) {
  const pad = " ".repeat(depth * indent);
  const label = key === null ? "" : fmtKey(key, d);

  if (guard > MAX_DEPTH) {
    lines.push(`${pad}${label}: <too deeply nested>`);
    return;
  }
  if (Array.isArray(value)) {
    writeArray(label, value, lines, pad, depth, d, indent, guard);
  } else if (value !== null && typeof value === "object") {
    writeObject(label, value, lines, pad, depth, d, indent, guard);
  } else {
    lines.push(`${pad}${label}${label ? ": " : ""}${fmtScalar(value, d)}`);
  }
}

function writeObject(label, obj, lines, pad, depth, d, indent, guard) {
  const keys = Object.keys(obj);
  if (!keys.length) {
    lines.push(label ? `${pad}${label}:` : pad);
    return;
  }

  // Keyed tabular: an object whose values are all uniform objects.
  const spec = keys.length > 1 ? tableSpec(keys.map((k) => obj[k])) : null;
  if (spec) {
    lines.push(`${pad}${label}[${keys.length}:]{${header(spec, d)}}:`);
    const inner = " ".repeat((depth + 1) * indent);
    for (const k of keys) {
      lines.push(`${inner}${fmtKey(k, d)}: ${cells(obj[k], spec, d).join(d)}`);
    }
    return;
  }

  let at = depth;
  if (label) {
    lines.push(`${pad}${label}:`);
    at += 1;
  }
  for (const k of keys) write(k, obj[k], lines, at, d, indent, guard + 1);
}

function writeArray(label, items, lines, pad, depth, d, indent, guard) {
  const n = items.length;
  if (!n) {
    lines.push(`${pad}${label}[0]:`);
    return;
  }
  if (items.every(isScalar)) {
    lines.push(`${pad}${label}[${n}]: ` + items.map((x) => fmtScalar(x, d)).join(d));
    return;
  }
  const spec = tableSpec(items);
  if (spec) {
    lines.push(`${pad}${label}[${n}]{${header(spec, d)}}:`);
    const inner = " ".repeat((depth + 1) * indent);
    for (const row of items) lines.push(inner + cells(row, spec, d).join(d));
    return;
  }

  lines.push(`${pad}${label}[${n}]:`);
  const lead = (depth + 1) * indent;
  for (const item of items) {
    const block = [];
    write(null, item, block, depth + 2, d, indent, guard + 1);
    if (!block.length) block.push(" ".repeat(lead + indent));
    const first = block[0];
    block[0] = (first.slice(0, lead) + "- " + first.slice(lead + indent)).replace(/\s+$/, "");
    lines.push(...block);
  }
}

// ---- deciding whether a table is possible ----

function tableSpec(rows) {
  if (!rows.length) return null;
  const plain = (v) => v !== null && typeof v === "object" && !Array.isArray(v);
  if (!rows.every((r) => plain(r) && Object.keys(r).length)) return null;

  const names = Object.keys(rows[0]);
  const wanted = new Set(names);
  for (const r of rows) {
    const ks = Object.keys(r);
    if (ks.length !== wanted.size || !ks.every((k) => wanted.has(k))) return null;
  }

  const spec = [];
  for (const name of names) {
    const column = rows.map((r) => r[name]);
    if (column.every(isScalar)) {
      spec.push([name, null]);
    } else if (column.every((v) => plain(v) && Object.keys(v).length)) {
      const sub = tableSpec(column);
      if (!sub) return null;
      spec.push([name, sub]);
    } else {
      return null;   // a list in a cell has nowhere to go on a flat row
    }
  }
  return spec;
}

function header(spec, d) {
  return spec
    .map(([name, sub]) => (sub ? `${fmtKey(name, d)}{${header(sub, d)}}` : fmtKey(name, d)))
    .join(d);
}

function cells(row, spec, d) {
  const out = [];
  for (const [name, sub] of spec) {
    if (sub) out.push(...cells(row[name], sub, d));
    else out.push(fmtScalar(row[name], d));
  }
  return out;
}

// ---- scalars ----

function isScalar(v) {
  return v === null || v === undefined || ["string", "number", "boolean"].includes(typeof v);
}

function fmtScalar(value, d) {
  if (value === null || value === undefined) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") {
    return Number.isFinite(value) ? String(value) : quote(String(value));
  }
  if (typeof value === "string") return needsQuote(value, d) ? quote(value) : value;
  return quote(String(value));
}

function needsQuote(s, d) {
  if (s === "" || s !== s.trim()) return true;
  if (s.includes(d) || s.includes('"') || CONTROL.test(s)) return true;
  if (LITERALS.has(s) || NUMBERISH.test(s)) return true;
  return "#-[{".includes(s[0]);
}

function quote(s) {
  let out = '"';
  for (const ch of s) {
    if (ESCAPES[ch]) out += ESCAPES[ch];
    else if (ch.charCodeAt(0) < 0x20 || ch.charCodeAt(0) === 0x7f) {
      out += "\\u" + ch.charCodeAt(0).toString(16).padStart(4, "0");
    } else out += ch;
  }
  return out + '"';
}

function fmtKey(key, d) {
  const s = String(key);
  if (!s || s !== s.trim()) return quote(s);
  if (s.includes(d) || KEY_BREAKERS.test(s) || CONTROL.test(s)) return quote(s);
  return s;
}

// ---- reading it ----

const ESC_HTML = { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" };
const escHtml = (s) => String(s).replace(/[&<>"]/g, (c) => ESC_HTML[c]);

// A header line: `name[3]{a,b}:` or `name[2:]{a,b}:` or `name[0]:`.
const HEAD = /^(\s*)([^[\]:]*)(\[\d+:?\])(\{.*\})?:(.*)$/;

/** TOON as highlighted HTML. Structure is coloured; values are not. */
export function highlightToon(text) {
  const out = [];
  for (const line of String(text).split("\n")) {
    const head = HEAD.exec(line);
    if (head) {
      const [, pad, name, count, fields, rest] = head;
      out.push(
        pad +
          `<span class="tn-key">${escHtml(name)}</span>` +
          `<span class="tn-count">${escHtml(count)}</span>` +
          (fields ? `<span class="tn-fields">${escHtml(fields)}</span>` : "") +
          `<span class="tn-punc">:</span>` +
          (rest ? `<span class="tn-row">${escHtml(rest)}</span>` : ""),
      );
      continue;
    }
    const kv = /^(\s*)(-\s)?([^:]+): (.*)$/.exec(line);
    if (kv) {
      const [, pad, dash, key, value] = kv;
      out.push(
        pad +
          (dash ? `<span class="tn-punc">${dash}</span>` : "") +
          `<span class="tn-key">${escHtml(key)}</span>` +
          `<span class="tn-punc">:</span> ` +
          `<span class="tn-val">${escHtml(value)}</span>`,
      );
      continue;
    }
    out.push(`<span class="tn-row">${escHtml(line)}</span>`);
  }
  return out.join("\n");
}

File: quickcode/context/toon.py
"""TOON: the encoding the model reads structured data in.

Token-Oriented Object Notation encodes the JSON data model, but declares a
uniform array's field list once in a header instead of repeating it on every
element::

    matches[3]{path,line,text}:
      src/a.py,12,def run():
      src/b.py,44,  run()
      src/c.py,91,"# run() is the entry point"

Two things are being bought here, and only one of them is tokens:

  - The header carries a row count. A model handed ``[200]`` that reads 198
    rows knows the answer was cut; with a bare list of lines it cannot tell
    truncation from an empty tail.
  - Fields are separated by a delimiter that values are *quoted against*. The
    old ``path:line:text`` rendering went ambiguous the moment a Windows path
    carried a drive letter, and nothing downstream could recover the split.

This is an encoder only. Nothing in QuickCode parses TOON back, because
nothing needs to: JSON stays the program's copy of the data and this is the
view handed to the model. A decoder we never call would be a second
implementation of the format to keep correct, for no reader.

Scope: model-facing text only. HTTP responses, session records and config
files stay JSON, because other programs read those.
"""

from __future__ import annotations

import math
import re
from typing import Any

# The spec allows comma, tab or pipe. Comma reads best and tokenizes well; tab
# is a little cheaper and is there for callers who know their values never
# contain one.
COMMA = ","
TAB = "\t"
PIPE = "|"

INDENT = 2

# Depth this far down is a cycle or a bug, and either way the model cannot use
# the result. Bail with a marker rather than blowing the stack.
MAX_DEPTH = 24

# Bare strings that would read back as something other than a string.
_LITERALS = frozenset({"true", "false", "null"})
_NUMBERISH = re.compile(r"^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$")
_CONTROL = re.compile(r"[\x00-\x1f\x7f]")

_ESCAPES = {
    "\\": "\\\\",
    '"': '\\"',
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}

# Characters that end a key or would be read as structure inside a header.
_KEY_BREAKERS = ':"{}[],'


def encode(value: Any, *, delimiter: str = COMMA, indent: int = INDENT) -> str:
    """Render ``value`` as TOON. Deterministic: same input, same bytes."""
    lines: list[str] = []
    _write(None, value, lines, 0, delimiter, indent, 0)
    return "\n".join(lines)


def fenced(value: Any, *, delimiter: str = COMMA) -> str:
    """``encode`` inside a ```toon fence.

    Worth the handful of extra tokens exactly once per result: naming the
    format is what stops a model reading a header line as prose.
    """
    return "```toon\n" + encode(value, delimiter=delimiter) + "\n```"


# ---- the four forms ----------------------------------------------------


def _write(
    key: str | None,
    value: Any,
    lines: list[str],
    depth: int,
    d: str,
    indent: int,
    guard: int,
) -> None:
    pad = " " * (depth * indent)
    label = "" if key is None else _fmt_key(key, d)

    if guard > MAX_DEPTH:
        lines.append(f"{pad}{label}: <too deeply nested>")
        return

    if isinstance(value, dict):
        _write_object(label, value, lines, pad, depth, d, indent, guard)
    elif isinstance(value, (list, tuple)):
        _write_array(label, list(value), lines, pad, depth, d, indent, guard)
    else:
        sep = ": " if label else ""
        lines.append(f"{pad}{label}{sep}{_fmt_scalar(value, d)}")


def _write_object(
    label: str,
    obj: dict,
    lines: list[str],
    pad: str,
    depth: int,
    d: str,
    indent: int,
    guard: int,
) -> None:
    if not obj:
        # Keyless (a list element) an empty object is a bare dash, so leave the
        # line empty here and let the caller put the dash on it.
        lines.append(f"{pad}{label}:" if label else pad)
        return

    # Keyed tabular: an object whose values are all uniform objects. Config
    # maps and records-by-id land here, and each row keeps its own key.
    spec = _table_spec(list(obj.values())) if len(obj) > 1 else None
    if spec is not None:
        lines.append(f"{pad}{label}[{len(obj)}:]{{{_header(spec, d)}}}:")
        inner = " " * ((depth + 1) * indent)
        for k, row in obj.items():
            lines.append(f"{inner}{_fmt_key(str(k), d)}: {d.join(_cells(row, spec, d))}")
        return

    if label:
        lines.append(f"{pad}{label}:")
        depth += 1
    for k, v in obj.items():
        _write(str(k), v, lines, depth, d, indent, guard + 1)


def _write_array(
    label: str,
    items: list,
    lines: list[str],
    pad: str,
    depth: int,
    d: str,
    indent: int,
    guard: int,
) -> None:
    n = len(items)
    if not items:
        lines.append(f"{pad}{label}[0]:")
        return

    # Inline form: primitives live on the header line.
    if all(_is_scalar(x) for x in items):
        lines.append(f"{pad}{label}[{n}]: " + d.join(_fmt_scalar(x, d) for x in items))
        return

    # Tabular form: the field list declared once, then one row per element.
    spec = _table_spec(items)
    if spec is not None:
        lines.append(f"{pad}{label}[{n}]{{{_header(spec, d)}}}:")
        inner = " " * ((depth + 1) * indent)
        for row in items:
            lines.append(inner + d.join(_cells(row, spec, d)))
        return

    # List form: the fallback that can hold anything.
    lines.append(f"{pad}{label}[{n}]:")
    lead = (depth + 1) * indent
    for item in items:
        # Rendered one level deeper than the dash, so an element's second and
        # later lines already sit under its first one.
        block: list[str] = []
        _write(None, item, block, depth + 2, d, indent, guard + 1)
        if not block:
            block = [" " * (lead + indent)]
        first = block[0]
        block[0] = (first[:lead] + "- " + first[lead + indent :]).rstrip()
        lines.extend(block)


# ---- deciding whether a table is possible ------------------------------

# A spec is a list of (field name, sub-spec or None). A sub-spec means the
# field holds uniform nested objects, which fold into the header as
# ``name{a,b}`` while the rows stay flat.
Spec = list[tuple[str, Any]]


def _table_spec(rows: list) -> Spec | None:
    """The field list for a tabular rendering, or None if the rows disagree."""
    if not rows or not all(isinstance(r, dict) and r for r in rows):
        return None
    names = [str(k) for k in rows[0]]
    wanted = set(names)
    if any({str(k) for k in r} != wanted for r in rows):
        return None
    if len(wanted) != len(names):  # two keys that stringify the same
        return None

    spec: Spec = []
    for name in names:
        column = [row[_key_named(row, name)] for row in rows]
        if all(_is_scalar(v) for v in column):
            spec.append((name, None))
        elif all(isinstance(v, dict) and v for v in column):
            sub = _table_spec(column)
            if sub is None:
                return None
            spec.append((name, sub))
        else:
            # A list in a cell has nowhere to go on a flat row.
            return None
    return spec


def _key_named(row: dict, name: str) -> Any:
    """The real key of ``row`` whose ``str()`` is ``name``."""
    if name in row:
        return name
    for k in row:
        if str(k) == name:
            return k
    raise KeyError(name)


def _header(spec: Spec, d: str) -> str:
    parts = []
    for name, sub in spec:
        key = _fmt_key(name, d)
        parts.append(key if sub is None else f"{key}{{{_header(sub, d)}}}")
    return d.join(parts)


def _cells(row: dict, spec: Spec, d: str) -> list[str]:
    out: list[str] = []
    for name, sub in spec:
        value = row[_key_named(row, name)]
        if sub is None:
            out.append(_fmt_scalar(value, d))
        else:
            out.extend(_cells(value, sub, d))
    return out


# ---- scalars -----------------------------------------------------------


def _is_scalar(value: Any) -> bool:
    return value is None or isinstance(value, (str, bool, int, float))


def _fmt_scalar(value: Any, d: str) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        # JSON has no spelling for these, so neither do we; saying so out loud
        # beats a bare `null` the model would read as "no value".
        if math.isnan(value) or math.isinf(value):
            return _quote(repr(value))
        return repr(value)
    if isinstance(value, str):
        return _quote(value) if _needs_quote(value, d) else value
    return _quote(str(value))


def _needs_quote(s: str, d: str) -> bool:
    if s == "" or s != s.strip():
        return True
    if d in s or '"' in s or _CONTROL.search(s):
        return True
    if s in _LITERALS or _NUMBERISH.match(s):
        return True
    # Leading punctuation that would read as structure rather than content.
    return s[0] in "#-[{"


def _quote(s: str) -> str:
    out = ['"']
    for ch in s:
        if ch in _ESCAPES:
            out.append(_ESCAPES[ch])
        elif ord(ch) < 0x20 or ord(ch) == 0x7F:
            out.append(f"\\u{ord(ch):04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _fmt_key(key: str, d: str) -> str:
    s = str(key)
    if not s or s != s.strip():
        return _quote(s)
    if d in s or any(c in s for c in _KEY_BREAKERS) or _CONTROL.search(s):
        return _quote(s)
    return s
